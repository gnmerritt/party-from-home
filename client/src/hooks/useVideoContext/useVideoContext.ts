import { useContext, useState, useEffect } from 'react';
import { VideoContext } from '../../components/VideoProvider';
import useMocks from '../../dev/useMocks';
import useParticipants from '../useParticipants/useParticipants';
import EventEmitter from 'events';
import { LocalTrackPublication } from 'twilio-video';

var useVideoContext = () => {
  const context = useContext(VideoContext);
  if (!context) {
    throw new Error('useVideoContext must be used within a VideoProvider');
  }
  return context;
};

if (process.env.REACT_APP_USE_MOCKS) {
  const fakeEmitter = new EventEmitter();

  useVideoContext = () => {
    const context = useContext(VideoContext);
    if (!context) {
      throw new Error('useVideoContext must be used within a VideoProvider');
    }
    const participants = useParticipants();
    const { connect, room, disconnect } = useMocks();
    const [contextRoom, setContextRoom] = useState<any>({
      ...context.room,
      state: 'disconnected',
      on: fakeEmitter.on,
      off: fakeEmitter.off,
      localParticipant: {
        identity: 'local',
        tracks: new Map(),
        on: fakeEmitter.on,
        off: fakeEmitter.off,
      },
    });

    useEffect(() => {
      const newRoom = {
        ...context.room,
        ...contextRoom,
      };

      if (room && (!contextRoom?.state || contextRoom?.state === 'disconnected')) {
        newRoom.state = 'connected';
        newRoom.name = room.id;
        newRoom.participants = participants;
        newRoom.disconnect = disconnect;
        const localVideoTrack: LocalTrackPublication = {
          sid: 'SR2323323',
          isSwitchedOff: false,
          setPriority: (_p: any) => {},
          isStarted: true,
          isTrackEnabled: true,
          name: 'camera',
          kind: 'video',
          trackName: 'camera',
          track: context.localTracks.find(t => t.kind === 'video'),
          on: (_e: any, _l: any) => {},
          off: (_e: any, _l: any) => {},
        } as any;
        const localAudioTrack: LocalTrackPublication = {
          sid: 'SR2323323',
          isSwitchedOff: false,
          setPriority: (_p: any) => {},
          isStarted: true,
          isTrackEnabled: true,
          name: 'audio',
          kind: 'audio',
          trackName: 'audio',
          track: context.localTracks.find(t => t.kind === 'audio'),
          on: (_e: any, _l: any) => {},
          off: (_e: any, _l: any) => {},
        } as any;
        newRoom.localParticipant = {
          ...contextRoom.localParticipant,
          sid: 'SR232321111',
          identity: 'local',
          tracks: [localVideoTrack, localAudioTrack],
        };
        setContextRoom(newRoom);
      } else if (!room && contextRoom?.state === 'connected') {
        newRoom.state = 'disconnected';
        setContextRoom(newRoom);
      }
    }, [context.localTracks, context.room, contextRoom, disconnect, participants, room]);

    return {
      ...context,
      room: contextRoom,
      isConnecting: false,
      connect,
    };
  };
}

export default useVideoContext;
