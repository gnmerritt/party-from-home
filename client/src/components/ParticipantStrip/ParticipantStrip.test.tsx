import EventEmitter from 'events';
import React from 'react';
import ParticipantStrip from './ParticipantStrip';
import { shallow } from 'enzyme';
import useSelectedParticipant from '../VideoProvider/useSelectedParticipant/useSelectedParticipant';
import useVideoContext from '../../hooks/useVideoContext/useVideoContext';

jest.mock('../../hooks/useVideoContext/useVideoContext');
jest.mock('../VideoProvider/useSelectedParticipant/useSelectedParticipant');
const mockedVideoContext = useVideoContext as jest.Mock<any>;
const mockUseSelectedParticipant = useSelectedParticipant as jest.Mock<any>;

describe('the ParticipantStrip component', () => {
  mockUseSelectedParticipant.mockImplementation(() => [null, () => {}]);

  it('should correctly render ParticipantInfo components', () => {
    const mockRoom: any = new EventEmitter();
    mockRoom.participants = new Map([
      [0, { sid: 0 }],
      [1, { sid: 1 }],
    ]);
    mockRoom.localParticipant = 'localParticipant';
    mockedVideoContext.mockImplementation(() => ({ room: mockRoom }));
    const wrapper = shallow(<ParticipantStrip />);
    expect(wrapper).toMatchSnapshot();
  });
});
