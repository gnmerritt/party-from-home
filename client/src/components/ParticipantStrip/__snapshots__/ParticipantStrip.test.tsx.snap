// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the ParticipantStrip component should correctly render ParticipantInfo components 1`] = `
<Styled(aside)>
  <Participant
    onClick={[Function]}
    participant="localParticipant"
  />
  <Participant
    key="0"
    onClick={[Function]}
    participant={
      Object {
        "sid": 0,
      }
    }
  />
  <Participant
    key="1"
    onClick={[Function]}
    participant={
      Object {
        "sid": 1,
      }
    }
  />
</Styled(aside)>
`;
